### PR TITLE
[gcb] Add support for ESLint+TypeScript

### PR DESCRIPTION
**Pull request by octogonz**
This replaces @iclanton's earlier PR https://github.com/microsoft/rushstack/pull/1565.

After thinking about this problem, I've decided to incorporate ESLint support directly into `gulp-core-build-typescript` and `rush-stack-compiler-*`, rather than introducing a new `gulp-core-build-eslint` package.  

Rationale:

- This will make it easier for projects to migrate from TSLint->ESLint; you simply replace `tslint.json` with `.eslintrc.js`.  No changes will be needed for `package.json` or configs.
- The extra footprint for always installing ESLint is negligible.  For example, an empty SPFx project today is `51,570 Files` / `419 MB`.  Adding ESLint and all plugins, it increases to `53,491 Files` / `430 MB` which is only about 3%
- We're planning to tackle these problems in a much more sophisticated way with the upcoming `heft` toolchain.  Thus the main goal with RSC is to quickly get everything using the new ESLint+Prettier ruleset, not to have a perfect implementation.

The plan:

After this PR is merged and published, we'll do another PR to convert all the projects to ESLint, then another PR to enable Prettier.

### CODE REVIEW

File: build-tests/node-library-build-tslint-test/.npmignore [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Ignore everything by default
+**
```
Comment (iclanton) [inline build-tests/node-library-build-tslint-test/.npmignore]: Probably should just ignore everything and mark this project as private.

File: core-build/gulp-core-build-typescript/src/LintCmdTask.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    super(
+      'lint',
+      {
+        displayAsError: false
+      }
+    );
+  }
+
+  public loadSchema(): Object {
+    return JsonFile.load(path.resolve(__dirname, 'schemas', 'tslint-cmd.schema.json'));
```
Comment (iclanton) [inline core-build/gulp-core-build-typescript/src/LintCmdTask.ts]: Probably should rename this file.

File: core-build/gulp-core-build-typescript/src/LintCmdTask.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  }
+
+  public executeTask(): Promise<void> {
+    this.initializeRushStackCompiler();
+
+    const tslintRunner: TLintRunner = new this._rushStackCompiler.LintRunner(
```
Comment (iclanton) [inline core-build/gulp-core-build-typescript/src/LintCmdTask.ts]: This variable should probably be called `lintRunner`

File: stack/rush-stack-compiler-shared/src/shared/LintRunner.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    super(taskOptions, rootPath, terminalProvider);
+
+    this._eslintRunner = new EslintRunner(taskOptions, rootPath, terminalProvider);
+    this._tslintRunner = new TslintRunner(taskOptions, rootPath, terminalProvider);
+  }
+
+  public invoke(): Promise<void> {
+    const tslintFilePath: string = path.join(this._standardBuildFolders.projectFolderPath, 'tslint.json');
+
+    let promise: Promise<void> = Promise.resolve();
```
Comment (iclanton) [inline stack/rush-stack-compiler-shared/src/shared/LintRunner.ts]: Just use `async`/`await`

File: stack/rush-stack-compiler-shared/src/shared/LintRunner.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    if (FileSystem.exists(tslintFilePath)) {
+      promise = promise.then(() => this._tslintRunner.invoke());
+    }
+
+    // ESLint supports a crazy amount of different filenames and formats for its config file.  To reduce
```
Comment (iclanton) [inline stack/rush-stack-compiler-shared/src/shared/LintRunner.ts]:
```suggestion
    // ESLint supports a crazy number of different filenames and formats for its config file.  To reduce
```

File: build-tests/node-library-build-tslint-test/.npmignore (hunk 1 above)
Comment (octogonz) [reply build-tests/node-library-build-tslint-test/.npmignore]: This is a good idea, but it applies to many projects.  We should handle it in a separate PR.

File: core-build/gulp-core-build-typescript/src/LintCmdTask.ts (hunk 2 above)
Comment (octogonz) [reply core-build/gulp-core-build-typescript/src/LintCmdTask.ts]: The schema is published on the website.  Not sure if anything is using it, so I copy+pasted and created a second file.

File: core-build/gulp-core-build-typescript/src/LintCmdTask.ts (hunk 3 above)
Comment (octogonz) [reply core-build/gulp-core-build-typescript/src/LintCmdTask.ts]: Fixed

File: stack/rush-stack-compiler-shared/src/shared/LintRunner.ts (hunk 4 above)
Comment (octogonz) [reply stack/rush-stack-compiler-shared/src/shared/LintRunner.ts]: Way better! 🤩

File: stack/rush-stack-compiler-shared/src/shared/LintRunner.ts (hunk 5 above)
Comment (octogonz) [reply stack/rush-stack-compiler-shared/src/shared/LintRunner.ts]: Fixed

**Comment by octogonz** [comment]
@patmill  @VesaJuvonen The TSLint project has been deprecated now that ESLint supports TypeScript. This PR makes it easy to switch over. You could announce this as a preview for the next SPFx release, and eventually make it the default for the generator.

**Comment by octogonz** [comment]
Here's the PR that migrates the projects to use ESLint: https://github.com/microsoft/rushstack/pull/1593

**Comment by msgx** [comment]
Seems like the [issue 4972](https://github.com/SharePoint/sp-dev-docs/issues/4972) from **sp-dev-docs** is about the same topic.

**Pull request merged**